### PR TITLE
refactor: move session object inside connector class

### PR DESCRIPTION
**Pull request by shekhar-rudder**
## Description of the change

- `Connector` class is now computing `self.session` in its `init` method
- Removed `session` argument from all the connector methods except for those which are called by Snowpark. Now `self.session` is being used. Had to do some hacky stuff to make it work for creating a `sproc`
- Extracted out common integration test to `utils` file

https://linear.app/rudderstack/issue/PRML-516/code-refactoring-make-session-object-a-member-of-connector-class

## Type of change
- [ ] Bug fix (non-breaking change that fixes an issue)
- [ ] New feature (non-breaking change that adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)

## Related issues

> Fix [#1]() 

### Development

- [ ] Lint rules pass locally
- [ ] The code changed/added as part of this pull request has been covered with tests
- [ ] All tests related to the changed code pass in development

### Code review 

- [ ]  This pull request has a descriptive title and information useful to a reviewer. There may be a screenshot or screencast attached
- [ ] "Ready for review" label attached to the PR and reviewers mentioned in a comment
- [ ] Changes have been reviewed by at least one other engineer
- [ ] Issue from task tracker has a link to this pull request

### CODE REVIEW

File: src/predictions/rudderstack_predictions/connectors/CommonWarehouseConnector.py [hunk 1]
```diff
@@ -811,32 +803,26 @@ def delete_local_data_folder(self) -> None:
...
     def build_session(self, credentials: dict):
         pass
 
     @abstractmethod
-    def run_query(self, session, query: str, response: bool) -> Optional[Sequence]:
+    def run_query(self, query: str, response: bool) -> Optional[Sequence]:
         pass
 
     @abstractmethod
-    def fetch_table_metadata(self, session, table_name: str) -> List:
```
Comment (dpatchigolla) [inline src/predictions/rudderstack_predictions/connectors/CommonWarehouseConnector.py:829]: Why is this removed?

File: src/predictions/rudderstack_predictions/connectors/SnowflakeConnector.py [hunk 2]
```diff
@@ -134,21 +127,24 @@ def check_table_entry_in_material_registry(
...
+        table = self.session.table(table_name)
         if filter_condition:
             table = self.filter_table(table, filter_condition)
         return table
 
     def get_table_as_dataframe(
         self, session: snowflake.snowpark.Session, table_name: str, **kwargs
     ) -> pd.DataFrame:
-        return self.get_table(session, table_name, **kwargs).toPandas()
+        # Duplicating some code to avoid passing session as argument
```
Comment (dpatchigolla) [inline src/predictions/rudderstack_predictions/connectors/SnowflakeConnector.py:142]: session is still being passed as an argument, and only for this function. what's the reason?

File: src/predictions/rudderstack_predictions/trainers/MLTrainer.py [hunk 3]
```diff
@@ -205,14 +205,14 @@ def select_best_model(self, models, train_x, train_y, val_x, val_y, models_map):
...
-    def prepare_label_table(self, connector: Connector, session, label_table_name: str):
+    def prepare_label_table(self, connector: Connector, label_table_name: str):
         pass
 
     @abstractmethod
     def plot_diagnostics(
         self,
         connector: Connector,
-        session,
+        session,  # session is being passed as argument since this function is called from Snowpark procedure
```
Comment (dpatchigolla) [inline src/predictions/rudderstack_predictions/trainers/MLTrainer.py:215]: There's connector object inside snowflake stored procedure too. That should also have a session object right?

**Review by dpatchigolla — APPROVED** [review]
Looks good. Wrote a few comments, mainly for clarifications.

File: src/predictions/rudderstack_predictions/connectors/CommonWarehouseConnector.py (hunk 1 above)
Comment (shekhar-rudder) [reply src/predictions/rudderstack_predictions/connectors/CommonWarehouseConnector.py:829]: Removed it since it is already defined as an abstract method in `Connector` class

File: src/predictions/rudderstack_predictions/connectors/SnowflakeConnector.py [hunk 4]
```diff
@@ -134,21 +127,24 @@ def check_table_entry_in_material_registry(
...
+        try:
+            session.sql(f"select * from {table_name} limit 1").collect()
+        except:
+            raise Exception(f"Table {table_name} does not exist or not authorized")
+        return session.table(table_name).toPandas()
```
[the remark this answers is not in the document]
Comment (shekhar-rudder) [reply src/predictions/rudderstack_predictions/connectors/SnowflakeConnector.py:147]: We can't use `self.get_table` as it is since internally the methods uses `self.session` which in Snowpark methods will be undefined

File: src/predictions/rudderstack_predictions/trainers/MLTrainer.py (hunk 3 above)
Comment (shekhar-rudder) [reply src/predictions/rudderstack_predictions/trainers/MLTrainer.py:215]: No, `connector.session` will be undefined in case of snowpark since we are resetting it [here](https://github.com/rudderlabs/rudderstack-profiles-classifier/pull/273/files#diff-3ec5ebbc25710f8388fadbecd251362dfa66b354951fe66b67ff2194fe52ce1eR138) before creating sproc

File: src/predictions/rudderstack_predictions/connectors/SnowflakeConnector.py (hunk 2 above)
Comment (shekhar-rudder) [reply src/predictions/rudderstack_predictions/connectors/SnowflakeConnector.py:142]: I meant instead of calling `self.get_table`, I duplicated that code here since `self.session` would be undefined inside `self.get_table`. To make it work, I would have to make `session` object an argument in `self.get_table`. I didn't want to do it since it would have meant modifying the signature of few more functions
